File: api/src/main/java/com/hearlers/gateway/presentation/http/v1/users/app/UserDtoMapper.java
package com.hearlers.gateway.presentation.http.v1.users.app;

import com.hearlers.api.proto.v1.model.User;
import com.hearlers.api.proto.v1.model.UserProfile;
import com.hearlers.api.proto.v1.service.UpdateUserRequest;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.NullValueCheckStrategy;
import org.mapstruct.ReportingPolicy;

@Mapper(
        unmappedTargetPolicy = ReportingPolicy.IGNORE,
        nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS
)
public interface UserDtoMapper {

    UserDto.User of(User user);
    UserDto.UserProfile of (UserProfile userProfile);

    @Mapping(target = "userId", source = "userId")
    UpdateUserRequest toUpdateUserRequest(String userId, UserDto.UpdateMyUserRequest request);

    UserDto.UpdateMyUserResponse toUpdateMyUserResponse(User user);

    UserDto.FindMyUserResponse toFindMyUserResponse(User user);
}

File: api/src/main/kotlin/com/hearlers/gateway/presentation/http/v1/user/app/mapper/UserDtoMapper.kt
package com.hearlers.gateway.presentation.http.v1.user.app.mapper

import com.hearlers.api.proto.v1.model.User
import com.hearlers.api.proto.v1.model.UserProfile
import com.hearlers.api.proto.v1.model.UserTracking
import com.hearlers.api.proto.v1.service.UpdateTrackingRequest
import com.hearlers.api.proto.v1.service.UpdateUserRequest
import com.hearlers.api.proto.v1.service.updateTrackingRequest
import com.hearlers.gateway.presentation.http.v1.user.app.dto.UserDto

object UserDtoMapper {

    fun of(user: User): UserDto.User {
        return UserDto.User(
            id = user.id,
            nickname = user.nickname,
            userProfile = of(user.userProfile)
        )
    }

    fun of(userProfile: UserProfile): UserDto.UserProfile {
        return UserDto.UserProfile(
            userProfile.profileImage,
            userProfile.gender,
            userProfile.mbti,
            userProfile.birthday
        )
    }

    fun of(userTracking: UserTracking): UserDto.UserTracking {
        return UserDto.UserTracking(
            userTracking.hasSeenIntroCutscene,
            userTracking.createdAt,
            userTracking.updatedAt,
            userTracking.deletedAt
        )
    }

    fun toUpdateUserRequest(userId: String, request: UserDto.UpdateMyUserRequest): UpdateUserRequest {
        val builder = UpdateUserRequest.newBuilder().setUserId(userId)
        request.nickname?.let { builder.setNickname(it) }
        request.profileImage?.let { builder.setProfileImage(it) }
        request.gender?.let { builder.setGender(it) }
        request.mbti?.let { builder.setMbti(it) }
        request.birthday?.let { builder.setBirthday(it) }
        return builder.build()
    }

    fun toUpsertTrackingRequest(userId: String, request: UserDto.UpsertMyTrackingRequest): UpdateTrackingRequest {
        return updateTrackingRequest {
            this.userId = userId
            this.hasSeenIntroCutscene = request.hasSeenIntroCutscene
        }
    }

    fun toUpdateMyUserResponse(user: User): UserDto.UpdateMyUserResponse {
        return UserDto.UpdateMyUserResponse(user = of(user))
    }

    fun toFindMyUserResponse(user: User): UserDto.FindMyUserResponse {
        return UserDto.FindMyUserResponse(user = of(user))
    }
}
